local-cli/generator/index.js

```diff
@@ -8,11 +8,41 @@
  */
 'use strict';
 
+var execSync = require('child_process').execSync;
 var fs = require('fs');
 var path = require('path');
+var semver = require('semver')
 var yeoman = require('yeoman-generator');
 var utils = require('../generator-utils');
 
+// Use Yarn if available, it's much faster than the npm client.
+// Return the version of yarn installed on the system, null if yarn is not available.
+function yarnVersionIfAvailable() {
+  let yarnVersion;
+  try {
+    if (process.platform.startsWith('win')) {
+      // TODO implement on Windows
+      return null;
+    } else {
+      // execSync returns a Buffer -> convert to string
+      yarnVersion = (execSync('yarn --version 2>/dev/null').toString() || '').trim();
+    }
+  } catch (error) {
+    return null;
+  }
+  // yarn < 0.16 has a 'missing manifest' bug
+  try {
+    if (semver.gte(yarnVersion, '0.16.0')) {
+      return yarnVersion;
```

Comment:
@bestander I don't love bundling Yarn because `react-native-cli` is supposed to not change very often and we put the changes in `react-native/local-cli` instead. As Yarn 0.17 and 0.18 etc. are released, we'll have to tell everyone to update `react-native-cli` too.

Also because of differences in the lockfile format and installation behavior between Yarn versions, it could be confusing if you already have Yarn installed but "rn init" creates a lockfile or node_modules that's different than what you're used to.

I suppose we could include yarn@^0.16 with r-n-cli and use it if the user doesn't have Yarn installed, but I do think it's good to first look for the user's own copy of Yarn. To me the question is: is it better to use `[user's own copy of Yarn -> fallback to included Yarn 0.16]` or is it better to use `[user's own copy of Yarn -> fallback to user's own copy of npm]`? There are good arguments to be made for both sides, on one hand we could focus on supporting just Yarn (especially for new projects), on the other using the user's own package manager might be less surprising.